src/docker/composeappengine.cc

```diff
@@ -12,28 +12,120 @@ ComposeAppEngine::ComposeAppEngine(boost::filesystem::path root_dir, std::string
     : root_{std::move(root_dir)},
       compose_{std::move(compose_bin)},
       docker_client_{std::move(docker_client)},
-      registry_client_{std::move(registry_client)} {
-  boost::filesystem::create_directories(root_);
-}
+      registry_client_{std::move(registry_client)} {}
 
 bool ComposeAppEngine::fetch(const App& app) {
-  boost::filesystem::create_directories(appRoot(app));
-  if (download(app)) {
-    LOG_INFO << "Validating compose file";
-    if (cmd_streaming(compose_ + "config", app)) {
-      LOG_INFO << "Pulling containers";
-      return cmd_streaming(compose_ + "pull --no-parallel", app);
+  boost::filesystem::create_directories(appRoot(app) / MetaDir);
+
+  bool result = false;
+
+  try {
+    AppState state(app, appRoot(app), true);
+    switch (state()) {
+      case AppState::State::kUnknown:
+      case AppState::State::kDownloadFailed: {
+        if (!download(app)) {
+          state.setState(AppState::State::kDownloadFailed);
+          break;
+        }
+        state.setState(AppState::State::kDownloaded);
+      }
+      case AppState::State::kDownloaded:
+      case AppState::State::kVerifyFailed: {
```

Comment:
if verify failed, shouldn't we try to fetch again?

Reply:
ah - i guess you need VerifyFailed in both places.

Reply:
What for?

Reply:
I assume verifyfailed could mean one of 2 things: Computing the hash failed (no sense retrying). The actual payload is wrong. In this case we should try and download it again, or the device will never apply the update?

Reply:
as discussed